packages/clay-modal/src/ClayModal.tsx

```diff
@@ -0,0 +1,206 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import Button from '@clayui/button';
+import classNames from 'classnames';
+import Icon from '@clayui/icon';
+import React, {FunctionComponent, useEffect, useRef, useState} from 'react';
+import {createPortal} from 'react-dom';
+import {Footer, FooterContainer, FooterContainerProps} from './Footer';
+import {useModal} from './Hook';
+
+type Size = 'full-screen' | 'lg' | 'sm';
+
+type Status = 'danger' | 'info' | 'success' | 'warning';
+
+interface Props extends React.HTMLAttributes<HTMLDivElement> {
+	/**
+	 * Callback called to close the modal.
+	 */
+	onClose: () => void;
+
+	/**
+	 * Renders the footer with the FooterContainer component utilities.
+	 */
+	renderFooter?: React.ReactElement;
+
+	/**
+	 * The size of element modal.
+	 */
+	size?: Size;
+
+	/**
+	 * The path to the SVG spritemap file containing the icons.
+	 */
+	spritemap?: string;
+
+	/**
+	 * Status messages.
+	 */
+	status?: Status;
+
+	/**
+	 * Title of the modal.
+	 */
+	title?: string;
+
+	/**
+	 * Url to place an iframe in the body of the modal.
+	 */
+	url?: string;
+
+	/**
+	 * Flag to indicate if the modal is open or closed.
+	 */
+	visible: boolean;
```

Comment:
By having this component manage its own visibility, we are now required to always render its contents right? I'm wondering if its best to always let the parent component manage the child's visibility. This is larger infra question as a whole though and one I am not quite sure I know the "right" answer. Normally I would always default to having parent components manage visibility and rendering, but idk if that is right.

Anyone else have any thoughts on this? @wincent @jbalsas @diegonvs

Reply:
Maybe this name can be a bit confusing, but who really manages the visibility of the component is Father, it defines through the click of its button to open or close the modal through the callback `onClose`, we need to have the modal content "visible" in order to create the animations in this case, such as adding the class `show` when visible is true.

I like this style to set the visibility of the component, I think we more freedom to do things within the component, such as animation.

Reply:
@jbalsas @wincent any thoughts on this? Just making sure this didn't get lost in the mix of GH notifications.

Reply:
Thanks @bryceosterhaus, I missed the original mention. I tend to agree with you that the parent should be responsible for choosing when to render. It is a bit strange to have something with `visible={false}` render something into the DOM. I don't think the modal should receive a prop from the outside; rather I would expect it to be: parent renders -> on mount, modal animates in -> on close, modal animates out, then calls parent -> parent unmounts dialog. Modal will obviously need internal state to manage the transitions: mounting/animating -> displayed -> closing/animating.

Reply:
hey @wincent the bad to go with the approach of animating the component in mount and unmount is that I have no control to suspend the unmount of the component until the animation finishes. Maybe I can control the onClose method to fire only when I finish the animation, but if I follow the approach I spoke in the other thread I will not have control over the onClick of the footer buttons.

Maybe we can accept a function in the `children` of `Modal` by passing the controlled method of onClose.

```jsx
<ClayModal onClose={onClose}>
 {handleClose => (
    <ClayModal.Footer
      primary={<ClayButton onClick={handleClose}>}
    />
  )}
</ClayModal>
```

What do you think?

Reply:
Yeah, I don't think you would want a way to "suspend" an unmount; the parent should be the one that controls when its children are mounted. That's why I said above that the modal would do the animation, then notify the parent, which actually does the unmounting. So what you say about calling `onClose` only at the end sounds like the right approach.

Having to pass through the `handleClose` callback probably isn't super ergonomic, but it could work. Did you look at prior art like [react-modal](https://github.com/reactjs/react-modal)? I see they have `onAfterOpen`, `onRequestClose` callbacks (and I also see that they do have an `isOpen` prop, although I haven't looked into the internals of it). I don't know anything about the project, so I can't vouch for its quality, but it might provide some food for thought.

Reply:
I took a look internally you can create an animation from their API, I think that's why they have many API's to customize the timeout, onAfterOpen to cover these cases... They have that power too due to the `isOpen` the even if we had `visible`, the component that controls when it will disassemble and mount.

I like the idea of ​​going with `handleClose`, I managed to make it work fine. I think it's something close to what Formik does with its handler functions, despite being another context and making more sense there.